src/Compilers/Core/Portable/Symbols/ISymbolExtensions_IsAccessibleWithin.cs

```diff
@@ -0,0 +1,353 @@
+            if (within == containingType)
+            {
+                return true;
+            }
+
+            // A nested symbol is only accessible to us if its container is accessible as well.
+            if (!IsAccessibleWithin(containingType, within, throughTypeOpt))
+            {
+                return false;
+            }
+
+            switch (declaredAccessibility)
+            {
+                case Accessibility.Public:
+                case Accessibility.NotApplicable:
+                    return true;
+
+                case Accessibility.Private:
+                    if (containingType.TypeKind == TypeKind.Submission)
+                    {
+                        return true;
+                    }
+
+                    return within != null && IsPrivateSymbolAccessible(within, containingType);
+            }
+
+            var withinType = within as INamedTypeSymbol;
+            var withinAssembly = withinType?.ContainingAssembly ?? (IAssemblySymbol)within;
+            switch (declaredAccessibility)
+            {
+                case Accessibility.Internal:
+                    return
+                        withinAssembly.HasInternalAccessTo(containingType.ContainingAssembly);
+
+                case Accessibility.ProtectedAndInternal:
+                    return
+                        IsProtectedSymbolAccessible(withinType, throughTypeOpt, containingType) &&
+                        withinAssembly.HasInternalAccessTo(containingType.ContainingAssembly);
+
+                case Accessibility.ProtectedOrInternal:
+                    return
+                        IsProtectedSymbolAccessible(withinType, throughTypeOpt, containingType) ||
+                        withinAssembly.HasInternalAccessTo(containingType.ContainingAssembly);
+
+                case Accessibility.Protected:
+                    return
+                        IsProtectedSymbolAccessible(withinType, throughTypeOpt, containingType);
+
+                default:
+                    throw ExceptionUtilities.UnexpectedValue(declaredAccessibility);
+            }
+        }
+
+        private static bool IsProtectedSymbolAccessible(INamedTypeSymbol withinType, ITypeSymbol throughTypeOpt, INamedTypeSymbol containingType)
+        {
+            INamedTypeSymbol originalContainingType = containingType.OriginalDefinition;
+            if (originalContainingType.TypeKind == TypeKind.Submission)
+            {
+                return true;
+            }
+
+            if (withinType == null)
+            {
+                // If we're not within a type, we can't access a protected symbol
+                return false;
+            }
+
+            if (withinType.IsNestedWithinOriginalContainingType(originalContainingType))
+            {
+                return true;
+            }
+
+            var originalThroughTypeOpt = throughTypeOpt == null ? null : throughTypeOpt.OriginalDefinition as ITypeSymbol;
```

Comment:
nit.  just use ```?.```?

Reply:
i'm not sure whyt the "as ITypeSymbol" is there.  .OriginalDefinition should already be an ITypeSymbol